### PR TITLE
Copy on write reshape

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88774

Design doc at https://docs.google.com/document/d/1kWIRUeixlgnOk1eXJSsbxOAKUKQFpmz0yzwLZUjWotg/edit#

This PR adds a copy kwarg to reshape, which triggers a copy-on-write reshape. The intention is to eventually make this default. It also adds a warning for improper use of mutation before/after reshape call.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @EikanWang @jgong5

### CODE REVIEW

**Review by thiagocrepaldi — APPROVED** [review]
torch/onnx/symbolic_*.py LGTM

File: torch/onnx/symbolic_helper.py [hunk 1]
```diff
@@ -1327,8 +1327,10 @@ def _index_fill_reshape_helper(g: jit_utils.GraphContext, self, dim, index):
 # the zero value is honored, similar to NumPy.
 # allowzero=1 is only supported for opset version >= 14.
 @_beartype.beartype
-def _reshape_helper(g: jit_utils.GraphContext, input, shape, allowzero=0):
+def _reshape_helper(g: jit_utils.GraphContext, input, shape, allowzero=0, copy=False):
     shape = _maybe_get_const(shape, "is")
+    copy = _maybe_get_const(copy, "b")
+    assert not copy, repr(copy)
```
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py:1333]: ONNX doesn't have concept of mutation. Should be fine to just ignore `copy`.

**Review by BowenBao — COMMENTED** [review]
My comment is only on ONNX part, inline suggestion and otherwise LGTM.

**Review by BowenBao — APPROVED** [review]

File: torch/onnx/symbolic_helper.py (hunk 1 above)
Comment (ezyang) [reply torch/onnx/symbolic_helper.py:1333]: fancy, that'll be great

**Comment by ezyang** [comment]
@dzhulgakov's idea is that we test if a storage is dirty on `data_ptr` access in storage; that will get us cheap reads, making the warnings more accurate.

**Pull request closed without merging**